src/java/org/apache/cassandra/index/sai/disk/v1/postings/MergePostingList.java

```diff
@@ -74,6 +74,10 @@ public long nextPosting() throws IOException
         // lazily create PQ if we haven't already
         if (pq == null)
         {
+            // elements could be removed in advance() even thouh postingLists started as non-empty
+            if (postingLists.isEmpty())
+                return PostingList.END_OF_STREAM;
```

Comment:
On second thought, I wonder if the contract isn't strictly followed. The `advance` method javadoc says:

```java
    /**
     * Advances to the first row ID beyond the current that is greater than or equal to the
     * target, and returns that row ID. Exhausts the iterator and returns {@link #END_OF_STREAM} if
     * the target is greater than the highest row ID.
     *
     * Note: Callers must use the return value of this method before calling {@link #nextPosting()}, as calling
     * that method will return the next posting, not the one to which we have just advanced.
     *
     * @param targetRowID target row ID to advance to
     *
     * @return first segment row ID which is >= the target row ID or {@link PostingList#END_OF_STREAM} if one does not exist
     */
    long advance(long targetRowID) throws IOException;
```

Reply:
The fix certainly is not wrong, but it seems like hitting this case requires that `lastRowId == targetRowID`, and I don't see why we would advance to a row id we've already seen. That seems suboptimal.

Reply:
I'm no longer sure that this is the right fix. It could be hiding an issue in the RangeIntersectionIterator where we try to skip to a given primary key in order to see if an iterator has that one primary key, but we accidentally go past it because the MPL returned the row id for that primary key in a previous call.

Reply:
I spent several hours looking into this with @dlg99, and we couldn't find a case where data was lost or the wrong results returned after this patch is applied.